src/Middleware/HostFiltering/src/MiddlewareConfigurationManager.cs

```diff
@@ -0,0 +1,120 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.ObjectModel;
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.HostFiltering;
+
+internal sealed record MiddlewareConfiguration(bool AllowAnyNonEmptyHost, bool AllowEmptyHosts, bool IncludeFailureMessage, ReadOnlyCollection<StringSegment>? AllowedHosts = default);
+
+internal sealed class MiddlewareConfigurationManager
+{
+    private readonly object _syncObject = new();
+    private event Func<HostFilteringOptions>? _getChangeRequestObject;
```

Comment:
I don't think this locking or indirection are needed. We just need to make sure _middlewareConfiguration is handled atomically.
```suggestion
```